File: Easy/121-Best-Time-To-Buy-And-Sell-Stock.java
// Time: O(N)
// Space: O(1)

class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int curMin = prices[0];
        
        for (int i = 0; i < prices.length; i++) {
            curMin = Math.min(curMin, prices[i]);
            profit = Math.max(profit, prices[i] - curMin);
        }
        
        return profit;
    }
}

File: Easy/121-Best-Time-To-Buy-And-Sell-Stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        curr_lowest, max_profit = prices[0], 0

        for i, v in enumerate(prices):
            if i == 0:
                continue
            
            max_profit = max(max_profit, v - curr_lowest)
            curr_lowest = min(curr_lowest, v)
                
        return max_profit
